### PR TITLE
Detect nesting `it` and `pending` at run-time

### LINKED ISSUE #7192: pending test causes incorrect test count and malformed output
Version of crystal:
```
Crystal 0.27.0 [c9d1eef8f] (2018-11-01)

LLVM: 4.0.0
Default target: x86_64-unknown-linux-gnu
```

When I mark a test as pending the count of tests and test output are malformed.

I wrote the following test.
```crystal
describe "my tests" do
  it "1 equals 1.0" do
    1.should eq 1.0
  end
end
```

Running the test correctly outputs one green test displays the expected output.
screen shot 2018-12-16 at 8 24 31 am

When I mark the test as pending, things will get interesting.
```crystal
describe "my tests" do
  it "1 equals 1.0" do
    pending "something"
    1.should eq 1.0
  end
end
```

The output shows two examples and the verbose output is malformed with the word `something` showing up twice on the first example listed and only `1.0` from the test name is shown, not the full name `1 equals 1.0`.
screen shot 2018-12-16 at 8 23 49 am

The non-verbose spec output also shows two examples, although is not malformed.
screen shot 2018-12-16 at 8 25 30 am

### PR DESCRIPTION
**Pull request by MakeNowJust**
Fix #7192
Close #7207

### CODE REVIEW

File: src/spec/context.cr [hunk 1]
```diff
@@ -166,6 +166,23 @@ module Spec
...
     end
+
+    @@spec_nesting = false
+
+    def self.check_nesting_spec(file, line)
+      check_nesting_spec(file, line) { }
+    end
+
+    def self.check_nesting_spec(file, line, &block)
+      raise NestingSpecError.new("cannot nest `it` and `pending`", file, line) if @@spec_nesting
```
Comment (Sija) [inline src/spec/context.cr]: `and` -> `or`

Comment (MakeNowJust) [reply src/spec/context.cr]: Thank you!

File: src/spec/expectations.cr [hunk 2]
```diff
@@ -385,6 +385,11 @@ module Spec
         raise ex
       end
 
+      # `NestingSpecError` is also treated.
+      if ex.is_a?(Spec::NestingSpecError) && klass != Spec::NestingSpecError
```
Comment (asterite) [inline src/spec/expectations.cr]: What is this for?

**Comment by MakeNowJust** [comment]
Please don't merge this for now because this doesn't fix #7192 entirely. Sorry 🙇 I'll fix it asap.

@straight-shoota At least we have an issue #7192 and this is the simplest way to fix it.

**Comment by MakeNowJust** [comment]
Now, nesting `pending`/`it` doesn't break `crystal spec -v` output.

Example:

```crystal
require "spec"

it "foo" do
  it "bar" do
  end
end

it "baz" do
  pending
end
```

Old (broken) output:



New (fixed) output:

File: src/spec/expectations.cr (hunk 2 above)
Comment (MakeNowJust) [reply src/spec/expectations.cr]: See full code:

```crystal
    def expect_raises(klass : T.class, message = nil, file = __FILE__, line = __LINE__) forall T
      yield
    rescue ex : T
      # We usually bubble Spec::AssertaionFailed, unless this is the expected exception
      if ex.is_a?(Spec::AssertionFailed) && klass != Spec::AssertionFailed
        raise ex
      end

      # `NestingSpecError` is treated as the same above.
      if ex.is_a?(Spec::NestingSpecError) && klass != Spec::NestingSpecError
        raise ex
      end

      # ...
    end
```

Please tell me more correct sentence which means this if you know.

File: src/spec/context.cr [hunk 3]
```diff
@@ -166,6 +166,19 @@ module Spec
     def matches?(pattern, line, locations)
       false
     end
+
+    @@spec_nesting = false
```
Comment (Fryguy) [inline src/spec/context.cr]: If we were to ever get to parallel test execution, would a class variable cause a problem?

Comment (asterite) [reply src/spec/context.cr]: Yes

Comment (asterite) [reply src/spec/context.cr]: Oh, actually, I was replying generically that class variables are to be used carefully in a parallel context. Not sure about this case, though

**Pull request merged**